src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonIncentiveLevelService.kt

```diff
@@ -167,6 +168,31 @@ class PrisonIncentiveLevelService(
     }
   }
 
+  /**
+   * Deactivate all incentive levels for a given prison; useful for when prisons are closed
+   */
+  @Transactional
+  suspend fun deactivateAllPrisonIncentiveLevels(prisonId: String): List<PrisonIncentiveLevelDTO> {
+    val prisonIncentiveLevels = prisonIncentiveLevelRepository.findAllByPrisonId(prisonId).toList()
+    val prisonIncentiveLevelsWithPrisoners = prisonIncentiveLevels.filter { prisonIncentiveLevel ->
+      countPrisonersService.prisonersExistOnLevelInPrison(prisonId, prisonIncentiveLevel.levelCode)
+    }
+    if (prisonIncentiveLevelsWithPrisoners.isNotEmpty()) {
+      throw ValidationExceptionWithErrorCode(
+        "A level must remain active if there are prisoners on it currently",
+        ErrorCode.PrisonIncentiveLevelActiveIfPrisonersExist,
+        moreInfo = prisonIncentiveLevelsWithPrisoners.map(PrisonIncentiveLevel::levelCode).joinToString(","),
+      )
+    }
+    prisonIncentiveLevelRepository.saveAll(
+      prisonIncentiveLevels.filter(PrisonIncentiveLevel::active)
+        .map { prisonIncentiveLevel ->
+          prisonIncentiveLevel.withUpdate(PrisonIncentiveLevelUpdateDTO(active = false))
+        },
+    ).collect()
+    return getAllPrisonIncentiveLevels(prisonId)
```

Comment:
Maybe if this returns the result of the `collect()` instead of re-reading the list of prison levels, it means `incentives.prison-level.changed` domain events would be published only for prison levels that were actually inactivated.

Reply:
ok, changed in eb64ca910eed378464c11acc390ae5953f1124be – now the service only returns the changed levels. the api endpoint still returns the full set given the caller wanted the _whole_ prison deactivated.